test/utils.dart

```diff
@@ -134,34 +134,52 @@ void _testFile(TestFile testFile) {
             indent: formatTest.leadingIndent,
             experimentFlags: const ['macros']);
 
-        var actual = formatter.formatSource(formatTest.input);
-
-        // The test files always put a newline at the end of the expectation.
-        // Statements from the formatter (correctly) don't have that, so add
-        // one to line up with the expected result.
-        var actualText = actual.text;
-        if (!testFile.isCompilationUnit) actualText += '\n';
-
-        // Fail with an explicit message because it's easier to read than
-        // the matcher output.
-        if (actualText != formatTest.output.text) {
-          fail('Formatting did not match expectation. Expected:\n'
-              '${formatTest.output.text}\nActual:\n$actualText');
-        } else if (actual.selectionStart != formatTest.output.selectionStart ||
-            actual.selectionLength != formatTest.output.selectionLength) {
-          fail('Selection did not match expectation. Expected:\n'
-              '${formatTest.output.textWithSelectionMarkers}\n'
-              'Actual:\n${actual.textWithSelectionMarkers}');
-        }
-
-        expect(actual.selectionStart, equals(formatTest.output.selectionStart));
-        expect(
-            actual.selectionLength, equals(formatTest.output.selectionLength));
+        var actual = _validateFormat(
+            formatter,
+            formatTest.input,
+            formatTest.output,
+            'did not match expectation',
+            testFile.isCompilationUnit);
+
+        // Make sure that formatting is idempotent. Format the output and make
+        // sure we get the same result.
+        _validateFormat(formatter, actual, formatTest.output,
+            'was not idempotent', testFile.isCompilationUnit);
       });
     }
   });
 }
 
+/// Run [formatter] on [input] and validate that the result matches [expected].
+///
+/// If not, fails with an error using [reason].
+///
+/// Returns the formatted output.
+SourceCode _validateFormat(DartFormatter formatter, SourceCode input,
+    SourceCode expected, String reason, bool isCompilationUnit) {
+  var actual = formatter.formatSource(input);
+
+  // The test files always put a newline at the end of the expectation.
+  // Statements from the formatter (correctly) don't have that, so add
+  // one to line up with the expected result.
+  var actualText = actual.text;
+  if (!isCompilationUnit) actualText += '\n';
+
+  // Fail with an explicit message because it's easier to read than
+  // the matcher output.
+  if (actualText != expected.text) {
+    fail('Formatting $reason. Expected:\n'
+        '${expected.text}\nActual:\n$actualText');
+  } else if (actual.selectionStart != expected.selectionStart ||
+      actual.selectionLength != expected.selectionLength) {
+    fail('Selection $reason. Expected:\n'
+        '${expected.textWithSelectionMarkers}\n'
+        'Actual:\n${actual.textWithSelectionMarkers}');
+  }
```

Comment:
I think using `expect` would give better output in some cases since it will point to the specific character of the first difference for the string equality test. What cases does this output have an advantage over `expect` with a `reason:` argument?

```suggestion
  expect(reason: 'Formatting $reason.', actualText, expected.text);
  expect(reason: 'Selection start $reason.', actual.selectionStart, expected.selectionStart);
  expect(reason: 'Selection length $reason.', actual.selectionLength, expected.selectionLength);
```

Reply:
It's really common for me to update a test expectation by simply copying the actual output from the failed test. Using `fail()` makes that easy: I just unindent the whole thing two spaces and it's done. The default output for strings turns them into string literals like:

```
    Actual: '  Widget build(BuildContext context) => CompositionRoot(\n'
              '    configureOverrides: configureOverrides,\n'
              '    compose:\n'
              '        (builder) =>\n'
              '            builder\n'
              '              //Adds a singleton CounterController to the container\n'
              '              ..addSingleton((container) => CounterController()),\n'
              '    child:\n'
              '    //We need the BuildContext from the Builder here so the children\n'
              '    //can access the container in the CompositionRoot\n'
              '    Builder(elided),\n'
              '    sdf,\n'
              '  );'
```

Which is much more annoying. Highlighting the first character difference is kind of nice, I admit. But in practice, it's usually pretty quick for me to spot the difference and being able to copy paste into tests is really handy.